File: webauthn-server-core/src/main/scala/com/yubico/webauthn/impl/KnownX509TrustAnchorsTrustResolver.scala
package com.yubico.webauthn.impl

import java.util.Optional

import com.fasterxml.jackson.databind.node.ArrayNode
import com.yubico.u2f.attestation.Attestation
import com.yubico.u2f.attestation.MetadataService
import com.yubico.u2f.data.messages.key.util.CertificateParser
import com.yubico.webauthn.AttestationTrustResolver
import com.yubico.webauthn.data.AttestationObject

import scala.collection.JavaConverters._


class KnownX509TrustAnchorsTrustResolver(
  private val metadataService: MetadataService
) extends AttestationTrustResolver {

  override def resolveTrustAnchor(attestationObject: AttestationObject): Optional[Attestation] =
    Optional.ofNullable(
      metadataService.getAttestation(
        attestationObject
          .attestationStatement
          .get("x5c")
          .asInstanceOf[ArrayNode]
          .iterator
          .asScala
          .map { node => CertificateParser.parseDer(node.binaryValue) }
          .toList
          .asJava
      )
    )

}

File: webauthn-server-core/src/main/java/com/yubico/webauthn/impl/KnownX509TrustAnchorsTrustResolver.java
package com.yubico.webauthn.impl;

import com.yubico.attestation.Attestation;
import com.yubico.attestation.MetadataService;
import com.yubico.util.CertificateParser;
import com.yubico.webauthn.AttestationTrustResolver;
import com.yubico.webauthn.data.AttestationObject;
import java.io.IOException;
import java.security.cert.CertificateException;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;


@Slf4j
@AllArgsConstructor
public class KnownX509TrustAnchorsTrustResolver implements AttestationTrustResolver {

    private final MetadataService metadataService;

    @Override
    public Optional<Attestation> resolveTrustAnchor(AttestationObject attestationObject) {
        return Optional.ofNullable(
            metadataService.getAttestation(
                StreamSupport.stream(
                    attestationObject
                        .getAttestationStatement()
                        .get("x5c")
                        .spliterator(),
                    true
                )
                    .map(node -> {
                        try {
                            return CertificateParser.parseDer(node.binaryValue());
                        } catch (CertificateException | IOException e) {
                            log.error("Failed to parse attestation certificate from attestation object: {}", attestationObject, e);
                            throw new RuntimeException(e);
                        }
                    })
                    .collect(Collectors.toList())
            )
        );
    }

}

